src/ErrorProne.NET/ErrorProne.NET.Test/ExceptionHandlingRules/ArgumentExceptionUsesStringAnalyzerTests.cs

```diff
@@ -0,0 +1,154 @@
+    }
+}";
+            HasDiagnostic(code, RuleIds.ArgumentExceptionParamNameRequired);
+        }
+
+        [Test]
+        public void ShouldWarnOnArgumentExceptionHasNoParametersInMethod() {
+            var code = @"
+using System;
+class Test
+{
+    public Foo(string arg) 
+    {
+        throw [|new ArgumentException()|];
+    }
+}";
+            HasDiagnostic(code, RuleIds.ArgumentExceptionParamNameRequired);
+        }
+
+        [Test]
+        public void ShouldWarnOnArgumentExceptionHasOnlyMessageParameterInCtor() {
+            var code = @"
+using System;
+class Test
+{
+    public Test(string arg) 
+    {
+        throw [|new ArgumentException(""It's an argument exception"")|];
+    }
+}";
+            HasDiagnostic(code, RuleIds.ArgumentExceptionParamNameRequired);
+        }
+
+        [Test]
+        public void ShouldWarnOnArgumentExceptionHasOnlyMessageParameterInMethod() {
+            var code = @"
+using System;
+class Test
+{
+    public Foo(string arg) 
+    {
+        throw [|new ArgumentException(""It's an argument exception"")|];
+    }
+}";
+            HasDiagnostic(code, RuleIds.ArgumentExceptionParamNameRequired);
+        }
+
+        [Test]
+        public void ShouldWarnOnArgumentExceptionHasWrongParameterNameInCtor() {
+            var code = @"
+using System;
+class Test
+{
+    public Test(string arg) 
+    {
+        throw new ArgumentException(""It's an argument exception"", [|""arg1""|]);
+    }
+}";
+            HasDiagnostic(code, RuleIds.ArgumentExceptionMethodHasNoSuchParamName);
+        }
+
+        [Test]
+        public void ShouldWarnOnArgumentExceptionHasWrongParameterNameInMethod() {
+            var code = @"
+using System;
+class Test
+{
+    public Foo(string arg) 
+    {
+        throw new ArgumentException(""It's an argument exception"", [|""arg1""|]);
+    }
+}";
+            HasDiagnostic(code, RuleIds.ArgumentExceptionMethodHasNoSuchParamName);
+        }
+
+        [Test]
+        public void ShouldWarnOnArgumentExceptionHasStringParameterNameInCtor() {
+            var code = @"
+using System;
+class Test
+{
+    public Test(string arg) 
+    {
+        throw new ArgumentException(""It's an argument exception"", [|""arg""|]);
+    }
+}";
+            HasDiagnostic(code, RuleIds.ArgumentExceptionParamNameShouldNotBeString);
+        }
+
+        [Test]
+        public void ShouldWarnOnArgumentExceptionHasStringParameterNameInMethod() {
+            var code = @"
+using System;
+class Test
+{
+    public Foo(string arg) 
+    {
+        throw new ArgumentException(""It's an argument exception"", [|""arg""|]);
+    }
+}";
+            HasDiagnostic(code, RuleIds.ArgumentExceptionParamNameShouldNotBeString);
+        }
+
+        [Test]
+        public void ShouldNotWarnOnArgumentExceptionHasNameofInCtor() {
+            var code = @"
+using System;
+class Test
+{
+    public Test(string arg) 
+    {
+        throw new ArgumentException(""It's an argument exception"", nameof(arg));
```

Comment:
> It means that here we need to cover ArgumentNullException that has paramName in different position and custom exception type.

No, I think there should be separate rules for `ArgumentNullException` and `ArgumentOutOfRangeException` - they have different set of constructors. Also, user can define its own class derived from `ArgumentException` that we can't analyze - unless we will add rules that will enforce him to implement this derived class in a specific way.

> I think implementation is missing other cases, when for paramName was constructed by something else except string literal. 

Yes, but I don't know what we can do in other cases. For example:
```{csharp}
public static void WrongState(string methodName, string argumentName, string state) {
    throw new ArgumentException($"It is not allowed to pass {argumentName} to method {methodName} when object is in {state} state", argumentName);
}
```
is completely valid helper, yet it doesn't follow our rule